client/grpc_query.go

```diff
@@ -24,100 +24,115 @@ var _ gogogrpc.ClientConn = Context{}
 var protoCodec = encoding.GetCodec(proto.Name)
 
 // Invoke implements the grpc ClientConn.Invoke method
-func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, args, reply interface{}, opts ...grpc.CallOption) (err error) {
+func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, req, reply interface{}, opts ...grpc.CallOption) (err error) {
 	// Two things can happen here:
 	// 1. either we're broadcasting a Tx, in which call we call Tendermint's broadcast endpoint directly,
 	// 2. or we are querying for state, in which case we call ABCI's Query.
 
-	// In both cases, we don't allow empty request args (it will panic unexpectedly).
-	if reflect.ValueOf(args).IsNil() {
+	// In both cases, we don't allow empty request req (it will panic unexpectedly).
+	if reflect.ValueOf(req).IsNil() {
 		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "request cannot be nil")
 	}
 
 	// Case 1. Broadcasting a Tx.
-	if isBroadcast(method) {
-		req, ok := args.(*tx.BroadcastTxRequest)
+	if IsGRPCBroadcastTx(method) {
+		reqProto, ok := req.(*tx.BroadcastTxRequest)
 		if !ok {
-			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "expected %T, got %T", (*tx.BroadcastTxRequest)(nil), args)
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "expected %T, got %T", (*tx.BroadcastTxRequest)(nil), req)
 		}
-		res, ok := reply.(*tx.BroadcastTxResponse)
+		resProto, ok := reply.(*tx.BroadcastTxResponse)
 		if !ok {
-			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "expected %T, got %T", (*tx.BroadcastTxResponse)(nil), args)
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "expected %T, got %T", (*tx.BroadcastTxResponse)(nil), req)
 		}
 
-		broadcastRes, err := TxServiceBroadcast(grpcCtx, ctx, req)
+		broadcastRes, err := TxServiceBroadcast(grpcCtx, ctx, reqProto)
 		if err != nil {
 			return err
 		}
-		*res = *broadcastRes
+		*resProto = *broadcastRes
 
 		return err
 	}
 
 	// Case 2. Querying state.
-	reqBz, err := protoCodec.Marshal(args)
+	inMd, _ := metadata.FromOutgoingContext(grpcCtx)
+	abciRes, outMd, err := RunGRPCQuery(ctx, grpcCtx, method, req, inMd)
 	if err != nil {
 		return err
 	}
 
+	err = protoCodec.Unmarshal(abciRes.Value, reply)
```

Comment:
Do you think it would be possible to store an object in the `abciRes` to avoid the serialization round trip? `RunGRPCQuery` would need to have additional parameter: `serialize bool` (or thous would be set in a header). If not set, we don't serialize, and instead set the value in the response directly.

Reply:
abciRes comes from tendermint, so not easily.

We could create a wrapper around it, for optimization (not serializing/deserializing the same structs multiple times...) as you say. But that's a bigger refactor, and if we're talking about bigger refactors, a more sensible one would be to allow concurrent gRPC queries.